From 084f33894395fc2e2fbc1f9b7aa7b6000bc705a3 Mon Sep 17 00:00:00 2001
From: lei shen <lei.shen@intel.com>
Date: Thu, 22 Oct 2015 12:08:37 +0800
Subject: [PATCH] SQUASHME! [VPG]: drm/i915: Update vblcount after disable
 maxfifo

Sometimes vblank is off when disable maxfifo, and vblcount
used is old value.So call drm_crtc_vblank_get to update
vblcount to latest before vblcount being used.

    SQUASHME! - This patch should be squashed into the following patch
        Author: vandita kulkarni <vandita.kulkarni@intel.com>
        Change-Id: I27bd9ed787558d0f59997225796aab04515a976e
        SQUASHME!: [VPG]: drm/i915: restructure maxfifo and add ddr dvfs

Change-Id: I09d99491476c5d81085098080a9489c8ef7d107b
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-9420
Signed-off-by: lei shen <lei.shen@intel.com>
Signed-off-by: Gajanan Bhat <gajanan.bhat@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 72aea74..6147378 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -11477,10 +11477,19 @@ static int intel_crtc_set_display(struct drm_crtc *crtc,
 				(pipe_stat & PIPE_ENABLE(PIPE_C) ||
 				(disp->update_flag &
 					DRM_MODE_SET_DISPLAY_UPDATE_ZORDER)))) {
+		int result;
 		intel_update_maxfifo(dev_priv, crtc, false);
 		dev_priv->wait_vbl = true;
+		/*
+		* Sometimes vblank is off at this time. Call
+		* drm_crtc_vblank_get to update vblcount to latest
+		* before vblcount is used, or wait_for_vblank may be missed.
+		*/
+		result = drm_crtc_vblank_get(crtc);
 		dev_priv->vblcount =
 			atomic_read(&dev->vblank[intel_crtc->pipe].count);
+		if (!result)
+			drm_crtc_vblank_put(crtc);
 	}
 
 
-- 
1.9.1

